### PR TITLE
Patch memory used for NeMo Megatron models

### PR DESCRIPTION
**Pull request by titu1994**
# What does this PR do ?

Reduce the peak memory usage during model building and restoration.

**Collection**: [NLP]

# Changelog 
- Force state dict onto CPU during restoration
- Add param_dtype in multiple places to correctly build tensor with trainer.precision dtype

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/attention.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
 from nemo.collections.nlp.modules.common.megatron.module import MegatronModule
 from nemo.collections.nlp.modules.common.megatron.rotary_pos_embedding import apply_rotary_pos_emb
 from nemo.collections.nlp.modules.common.megatron.utils import ApexGuardDefaults, attention_mask_func
+from nemo.collections.nlp.parts import utils_funcs as utils
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/attention.py:30]: Can we do this import without aliasing?

File: nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py [hunk 2]
```diff
@@ -24,6 +24,7 @@
     parallel_lm_logits,
     scaled_init_method_normal,
 )
+from nemo.collections.nlp.parts import utils_funcs as utils
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py:27]: can we do this import without aliasing?

File: nemo/collections/nlp/modules/common/megatron/language_model.py [hunk 3]
```diff
@@ -29,6 +29,7 @@
     init_method_normal,
     scaled_init_method_normal,
 )
+from nemo.collections.nlp.parts import utils_funcs as utils
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/language_model.py:32]: can we do this import without aliasing?

File: nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py [hunk 4]
```diff
@@ -26,6 +26,7 @@
     init_method_normal,
     scaled_init_method_normal,
 )
+from nemo.collections.nlp.parts import utils_funcs as utils
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py:29]: can we do this import without aliasing?

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 5]
```diff
@@ -38,6 +38,7 @@
     scaled_init_method_normal,
 )
 from nemo.collections.nlp.modules.common.megatron.vocab_parallel_cross_entropy import vocab_parallel_cross_entropy
+from nemo.collections.nlp.parts import utils_funcs as utils
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:41]: can we do this import without aliasing?

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 6]
```diff
@@ -38,6 +38,7 @@
 from nemo.collections.nlp.modules.common.megatron.mlp import ParallelMLP, SwitchMLP
 from nemo.collections.nlp.modules.common.megatron.module import MegatronModule
 from nemo.collections.nlp.modules.common.megatron.utils import ApexGuardDefaults
+from nemo.collections.nlp.parts import utils_funcs as utils
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:41]: can we do this import without aliasing?

File: nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py:27]: Sure but this is nlp utils module so it seems better to alias it. For example, if it was utils from other folders sure id use the full name but for the domains main utility module it seems excessive. This module should originally have been named just utils, i didn't change it cause its entrenched in multiple files already.

Comment (titu1994) [reply nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py]: Done

File: nemo/collections/nlp/modules/common/megatron/attention.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/nlp/modules/common/megatron/attention.py]: Done

File: nemo/collections/nlp/modules/common/megatron/language_model.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/nlp/modules/common/megatron/language_model.py]: Done

File: nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py]: Done

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py]: Done

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 6 above)
Comment (titu1994) [reply nemo/collections/nlp/modules/common/megatron/transformer.py]: Done

**Review by titu1994 — COMMENTED** [review]
Address comments

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**